patroni/ha.py

```diff
@@ -579,6 +626,30 @@ def post_recover(self):
             return 'failed to start postgres'
         return None
 
+    def handle_starting_instance(self):
+        """Starting up PostgreSQL may take a long time. In case we are the leader we may want to
+        fail over to."""
+        # state_handler.state == 'starting' here
+        if self.has_lock():
+            time_left = self.patroni.config['master_start_timeout'] - self.state_handler.time_in_state()
+            if time_left <= 0:
+                if self.is_failover_possible(self.cluster.members):
+                    logger.info("Demoting self because master startup is taking too long")
+                    self.demote_asap()
+                    return 'stopped PostgreSQL because of startup timeout'
+                else:
+                    return 'master start has timed out, but continuing to wait because failover is not possible'
+            else:
+                msg = self.process_manual_failover_from_leader()
+                if msg is not None:
+                    return msg
+
+                return 'PostgreSQL is still starting up, {0:.0f} seconds until timeout'.format(time_left)
+        else:
+            # Use normal processing for standbys
+            logger.info("Still starting up as a standby.")
+            return None
```

Comment:
Deferring to the main cycle handles:

* Retry starting PostgreSQL when it goes down.
* Changing primary_conninfo when leader changes
* Scheduled restarts (not sure if that is desired or not)

The main motivation was actually handling reinit, since I wrote this before 6dc1d9c88eae163ece5292fa4b8696f99bf9e754, but that doesn't seem to be relevant any more.